text/0000-cargo-workspace-deduplicate.md

```diff
@@ -0,0 +1,633 @@
+where you have to interpret multiple manifests to figure out what one means. For
+example you can no longer quickly and easily be guaranteed to parse the version
+of a package, but you might have to go find the workspace root or other crates
+to figure that out. Workspace discovery and membership is pretty nontrivial so
+non-Cargo based tools will have a difficult time *not* using Cargo to figure out
+a full elaborated form of a manifest 100% of the time.
+
+This proposal also extends `Cargo.toml` with changes that will break any
+existing tools which assume a particular format of `Cargo.toml`. For example if
+a tool expects `package.vesion` to be a `String` that runs a risk of being
+broken in the future due to the ability to specify a table there instead.
+
+Additionally this proposal complicates a reader's understanding of `Cargo.toml`.
+While verbose for maintainers having duplication of information is actually
+quite nice for readers of `Cargo.toml` because you don't have to chase anything
+else down to figure out what a dependency is. If this proposal is implemented
+then whenever you see `foo = { workspace = true }` you've got to go consult
+something else to figure out what the dependency actually is. This layer of
+indirection can cause surprise for readers or otherwise add a speed-bump to
+understanding the contents of a manifest.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Cargo's manifests have been a pretty carefully curated part of Cargo's design to
+ensure that they're consistently readable and concise where possible. For
+example many of Cargo's manifest idioms gently nudge users towards the same
+standards across the community by supporting many zero-configuration situations
+such as where to put and how to name tests.
+
+This proposal is an extension of these design principles to provide a gentle
+nudge to consistently, across the Rust community, manage workspaces,
+dependencies, and metadata. A goal here is to increase consistency in how this
+is all managed across projects in a way that still preserves Cargo's existing
+flexibility for users.
+
+Note that flexibility is a key part of this proposal where it's possible to
+intermingle shorthands with longer versions. For example if the workspace
+declares:
+
+```toml
+[workspace.dependencies]
+log = "0.3"
+```
+
+But you really want to try out a new version of `log` in one workspace member,
+you can easily do so by changing
+
+```toml
+[dependencies]
+log = { workspace = true }
+```
+
+to
+
+```toml
+[dependencies]
+log = "0.4"
+```
+
+Additionally you can always custom-version your packages, you've just got the
+option to reference another package as well. Overall this proposal should
+empower more power users of Cargo to manage workspaces easily without taking
+away any of the existing configurations that Cargo already supports.
+
+## Alternative Syntax
```

Comment:
Is the `workspace = true` entry in a dependency table actually necessary? It seems like that could be inferred if the dependency had no `version`, `git`, or `path` entry, right? So the log case would just look like

```toml
[dependencies]
log = {}
```

Reply:
I don't think we have a current technical reason requiring *something* to be here in that we could likely assign `workspace = true` meaning to an empty table. That being said though I would personally find the `log = {}` a bit too "small" in that it's sort of weird seeing no other indication of where the dependency comes from. I'd hope that `workspace = true` wouldn't be too burdensome though to list on lots of repeated dependencies.

Put another way, I think that this is a technically viable option, but I would say that for learnability and skimming crates it'd be nice to have something listed there to remind readers to look at the workspace manifest. I'm curious what others think though!

Reply:
Seems like the common case is going to be
```
[dependencies]
foo1 = { workspace = true }
foo2 = { workspace = true }
foo3 = { workspace = true }
foo4 = { workspace = true }
foo5 = { workspace = true }
...
```
with all or almost all crates being redirected to the workspace.

Which kind of asks for refactoring it into something looking like
```
[dependencies.workspace]
foo1
foo2
foo3
foo4
foo5
...
```
(but with valid syntax etc.)

Reply:
One of the more compact ways I've thought of is 

```toml
[dependencies]
foo = "ws"
```

Reply:
or maybe just

```toml
[dependencies]
foo1 = {}
foo2 = {}
```

Reply:
Note that the main syntax can also be written

```toml
[dependencies]
foo.workspace = true
```

Reply:
I like the `foo = "ws"`.

Reply:
I like the `foo = "workspace"`. Elaborate and concise at the same time.

Reply:
> [dependencies]
> foo1 = { workspace = true }
> foo2 = { workspace = true }
> foo3 = { workspace = true }
> foo4 = { workspace = true }
> foo5 = { workspace = true }

I disagree, I think this allows for future updates to this features? How about supporting something like 

> [dependencies]
> foo1, foo2, foo3, foo4, foo5 = { workspace = true }

Reply:
Is that valid toml?